Reading from files in mknetid was not handled properly and files with either
empty lines or unexpected format could cuase segmentation faults.

Steps to reproduce:
1. echo -e "\nsome incomplete entry\njoe:x:500:500:Joe:/home/joe:/bin/bash">fakepasswd
2. /usr/lib64/yp/mknetid -p fakepasswd

RHBZ: #1071962

This patch is back-ported from ypserv-2.32.

--- ypserv-2.19-orig-6.6/mknetid/mknetid.c	2014-04-25 13:45:49.346719623 +0200
+++ ypserv-2.19/mknetid/mknetid.c	2014-04-25 14:12:47.327733590 +0200
@@ -171,11 +171,30 @@ main (int argc, char *argv[])
 	{
 
 	  char *ptr, *key, *uid, *gid;
+	  char *err_line = strdup (line);
 
 	  key = xstrtok (line, ':');
 	  ptr = xstrtok (NULL, ':');
 	  uid = xstrtok (NULL, ':');
 	  gid = xstrtok (NULL, ':');
+
+	  if (key == NULL || ptr == NULL ||
+	      uid == NULL || gid == NULL)
+	    {
+	      int n = strlen (err_line);
+
+	      if (err_line[n - 1] == '\n')
+		err_line[n - 1] = '\0';
+
+	      if (strlen (err_line) > 0)
+		fprintf (stderr, "WARNING: bad netid entry: '%s'\n",
+			 err_line);
+	      free (err_line);
+	      continue;
+	    }
+
+	  free (err_line);
+
 	  if (insert_user (key, domain, uid, gid) < 0)
 	    if (!quiet_flag)
 	      fprintf (stderr, "WARNING: unix.%s@%s multiply defined, ignore new one\n",
@@ -200,21 +219,40 @@ main (int argc, char *argv[])
       if (n < 1)
 	break;
 
+      if (line[n - 1] == '\n')
+	{
+	  if(n < 2)
+	    continue; /* empty line */
+	  line[n - 1] = '\0';
+	}
+
       if (line[0] != '+' && line[0] != '-')
 	{
 	  char *grpname, *ptr, *gid, *user;
 
-	  if (line[strlen (line) - 1] == '\n')
-	    line[strlen (line) - 1] = '\0';
-
 	  grpname = xstrtok (line, ':');
 	  ptr = xstrtok (NULL, ':');
 	  gid = xstrtok (NULL, ':');
-	  while ((user = xstrtok (NULL, ',')) != NULL)
+	  if(grpname && ptr && gid)
+	    length -= gid - line;
+	  else
+	    {
+	      fprintf (stderr, "ERROR: bad format of group \"%s\".\n",
+		       grpname);
+	      exit (1);
+	    }
+	  user = xstrtok (NULL, ',');
+	  while (user != NULL) {
+	    length = length - strlen(user) - 1;
 	    if (add_group (user, gid) < 0)
 	      if (!quiet_flag)
 		fprintf (stderr, "WARNING: unknown user \"%s\" in group \"%s\".\n",
 			 user, grpname);
+	    if (length > 0)
+	      user = xstrtok (user + strlen(user) + 1, ',');
+	    else
+	      user = NULL;
+	  }
 	}
     }
   fclose (file);
@@ -273,8 +311,17 @@ main (int argc, char *argv[])
 
 	  if (line[0] != '#')
 	    {
+	      int n = strlen (line);
+
+	      if (line[n - 1] == '\n')
+		line[n - 1] = '\0';
 	      if (strpbrk (line, " \t") == NULL)
-		fprintf (stderr, "WARNING: bad netid entry: '%s'", line);
+		{
+		  if (strlen (line) > 0)
+		    fprintf (stderr,
+			     "WARNING: bad netid entry: '%s'\n",
+			     line);
+		}
 	      else
 		printf ("%s\n", line);
 	    }
